### ISSUE TITLE
(lambda-python): handle common dependencies

### ISSUE THREAD

**Issue by erudisch** [aws/aws-cdk#11837]
This issue is about common dependencies in python lambdas. For dealing with python lambdas there is the [pythonFunction](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-lambda-python-readme.html) available.

My lambda structure looks as follows:

```
/lambdas
  /common
    /common
      __init__.py
      common.py
    setup.py
  /lambda1
    lambda1.py
  /lambda2
    lambda2.py
```

`common.py` has some dependencies, f.ex. [stringutils](https://pypi.org/project/stringutils/). `lambda1` and `lambda2` needs functionality from common.py.

I want to be able to use `common.py` locally and inside `lambda1` and `lambda2` when they are deployed.

To achieve this my idea was to create a local package out of `common`. This `common` package should include the source code and the needed dependencies. Then I cloud just `import common` and use it's methods like an external library.

To install it locally I run `pip install -e .` from the `common` folder - this works fine.

But how can I provide my deployed lambdas with this package?

My idea was to run `python3 setup.py sdist --formats zip` to create a distribution package and use this zip file as entry for my `PythonLayerVersion` so that deployed `lambda1` and `lambda2` can use this module.

The creation of the `common` local package should also be handled by the pipeline. But when I run `python3 setup.py sdist --formats zip` when creating the lambda layer the egg-info top_level.txt doesn't include the correct path of `common`.

At this point I don't know how to proceed or if this way is correct.

To summarise everything:

How to provide common packages to lambdas?
These packages are self build and include other 3rd party packages.
The common packages should work locally and on the deployed lambda.

Thanks in advance
Erik

This is a 📕 documentation issue

**Comment by ryanjdillon** [comment]
I achieved this by doing the follwing:

In CDK
```typescript
import {PythonFunction, PythonLayerVersion} from '@aws-cdk/aws-lambda-python';

// CDK stack signature etc., then... 

    // AWS helpers lambda layer
    const helperLambdaLayer = new PythonLayerVersion(this, 'HelperLayer', {
      entry: './python-src/awshelpers',
      compatibleRuntimes: [lambda.Runtime.PYTHON_3_7],
    });

    // The function that uses the layer
    const myFunction =
        new PythonFunction(this, 'MyFunction', {
          entry : './python-src/lambdas/myfunction',
          handler : 'lambda_handler',
          runtime : lambda.Runtime.PYTHON_3_7,
          layers: [helperLambdaLayer],
        });
  // rest of your CDK app 
```

Where my helper module is setup like this:

```
cdkapp-root/python-src/
├── awshelpers
│   ├── awshelpers
│   │   ├── __init__.py
│   │   └── helpers.py
│   └── setup.py
└── lambdas
     └── mymodule
          ├── README.md
          ├── index.py
          ├── poetry.lock
          ├── pyproject.toml
          └── test_index.py
          └── treeid
```
And in my lambda's `index.py`:
```python
from awshelpers.helpers import my_func
```

**Comment by corymhall** [comment]
@erudisch does the solution provided by ryanjdillon work for you?

**Comment by kornicameister** [comment]
This is a way to go for me as well.
Common code was always put into a layer and wherever needed layer was added to lambda.
On development level I've been utilizing dev dependencies to install common code into virtual env.
